src/io/trailer.cc

```diff
@@ -0,0 +1,156 @@
+    trailer_ptr->SetVersion(version.VersionNumber());
+    trailer_ptr->SetIndexName(name);
+    writer.write(trailer_ptr->bytes, KNOWHERE_TRAILER_SIZE);
+}
+
+bool
+CheckTrailerForMemoryIO(MemoryIOReader& reader, const std::string& name) {
+    uint64_t bin_size = TRAILER_OFFSET(reader.size());
+    if (bin_size < 0) {
+        LOG_KNOWHERE_WARNING_ << "The binary is too small and assume no Trailer, pass Trailer check.";
+        return true;
+    }
+
+    auto trailer_ptr = std::make_unique<Trailer>();
+    auto pre_rp = reader.tellg();
+    reader.seekg(bin_size);
+    reader.read(trailer_ptr.get(), KNOWHERE_TRAILER_SIZE);
+    reader.seekg(pre_rp);
+
+    if (!trailer_ptr->TrailerValidCheck()) {
+        LOG_KNOWHERE_WARNING_ << "Trailer not exist in Binary.";
+        return true;
+    }
+
+    auto version = Version(trailer_ptr->GetVersion());
+    if (!Version::VersionSupport(version)) {
+        LOG_KNOWHERE_ERROR_ << "Index version(" << version.VersionNumber() << ") is not supported, Trailer check fail.";
+        return false;
+    }
+
+    if (trailer_ptr->GetIndexName() != name) {
+        LOG_KNOWHERE_ERROR_ << "Index type or data type is not correct(" << name << ").";
+        return false;
+    }
+
+    if (trailer_ptr->GetIndexBinarySize() != bin_size) {
+        LOG_KNOWHERE_ERROR_ << "The size of index binary is not correct.";
+        return false;
+    }
+
+    if (CalculateCheckSum(reader.data(), bin_size) != trailer_ptr->GetCheckSum()) {
+        LOG_KNOWHERE_ERROR_ << "Binary checksum check fail.";
+        return false;
+    }
+    LOG_KNOWHERE_INFO_ << "Index Trailer check succeed.";
+    return true;
+}
+
+void
+AddTrailerForFiles(const std::vector<std::string>& files, const std::string& trailer_file, const std::string& name,
+                   const Version& version) {
+    auto trailer_ptr = std::make_unique<Trailer>();
+    trailer_ptr->SetCheckSum(GetFilesCheckSum(files));
+    trailer_ptr->SetVersion(version.VersionNumber());
+    trailer_ptr->SetIndexName(name);
+
+    std::ofstream writer(trailer_file.c_str(), std::ios::binary);
+    writer.write((char*)trailer_ptr->bytes, KNOWHERE_TRAILER_SIZE);
+    writer.close();
+}
+
+bool
+CheckTrailerForFiles(const std::vector<std::string>& files, const std::string& trailer_file, const std::string& name) {
+    std::ifstream reader(trailer_file.c_str(), std::ios::binary);
+    if (!reader) {
+        LOG_KNOWHERE_WARNING_ << "Trailer file not exist.";
+        return true;
+    }
+    reader.seekg(0, std::ios::end);
+    auto fsize = reader.tellg();
+
+    reader.seekg(0, std::ios::beg);
+    if (fsize != KNOWHERE_TRAILER_SIZE) {
+        LOG_KNOWHERE_ERROR_ << "Trailer size (" << fsize << ")not correct.";
+        return false;
+    }
+    auto trailer_ptr = std::make_unique<Trailer>();
+    reader.read((char*)trailer_ptr.get()->bytes, KNOWHERE_TRAILER_SIZE);
+    if (!trailer_ptr->TrailerValidCheck()) {
```

Comment:
avoid duplicating those checks in CheckTrailerForFiles and CheckTrailerForMemoryIO. trailer class can have method `validate(name, checksum)` to do those checks.

Reply:
next pr. This pr verifies old knowhere code also can deserialize with the new binaryset.